### ISSUE TITLE
Can't preview a specific domain with OVH provider

### ISSUE THREAD

**Issue by tiennou** [StackExchange/dnscontrol#412]
I stumbled on the `--domains` argument of `preview`, which could be helpful in keeping the output sane, but it seem to be causing some kind of failure when using the OVH provider.

I'm getting this output with `dnscontrol preview --domains one-of-my.domains` :
```
******************** Domain: one-of-my.domains
----- DNS Provider: ovh...ERROR
Error getting corrections: Domain one-of-my.domains not found in your ovh account
Done. 0 corrections.
Completed with errors
```

While a normal `dnscontrol preview` results in the following :
```
[gobs of output for all domains]
******************** Domain: one-of-my.domains
----- DNS Provider: ovh...21 corrections
[21 DNS corrections]
----- Registrar: ovh...1 correction
#1: Change Nameservers from 'dns102.ovh.net,ns102.ovh.net' to 'ns1.one-of-my.domains,ns2.one-of-my.domains'
Done. 153 corrections.
Completed with errors
```

**Comment by captncraig** [comment]
That seems... odd. Are you running the exact same commands except for `push` and `preview`? Its like the same code... Hmm.

**Comment by tiennou** [comment]
It seems to be pretty consistent. I just noticed that I had some not-enabled zones in OVH (which is a known caveat that they have to be manually enabled), but even after enabling all zones, it still fails when a specific domain is "selected".

**Comment by tiennou** [comment]
Also, `push` errors in the same way (though I tested against a not-changed zone).

**Comment by tlimoncelli** [comment]
OVH is a community-supported provider.  Maybe @masterzen can look into this?

**Comment by masterzen** [comment]
Indeed, I'm going to look at it @tlimoncelli.

**Comment by masterzen** [comment]
@tiennou , I'm sorry I was unable to reproduce the issue with the latest master. I have 2 test domains out of 5 defined in my `dnsconfig.js`. Those domains have  items to modify at ovh, and `--domains` is working correctly, selecting the correct domains to either preview or apply.

What version of `dnscontrol` did you use?
What could be different in your setup?

One possibility is that your consummer key authorizations doesn't allow this domain to be modified or listed.

Looking at the ovh provider code, this error is raised when the `--domains` domain is not listed in the zone list accessible by your account as returned by the [list zone API call](https://api.ovh.com/console/#/domain/zone#GET).
Can you manually try this call in the ovh api console (see the previous link)  and check what it returns?

**Comment by tlimoncelli** [comment]
The `--domains` handler isn't part of the providers. It is a core part of the system.  Is there a chance of a typo?  domains vs domains? `--` vs `-`?  The `--domains` flag must be after the "preview" command.

Are you using the "master" branch or the last release?

Could you paste the exact command line you are using?

Tom

**Comment by tiennou** [comment]
I'm running #377 (which I just rebased on top of master to be extra sure, and shouldn't matter IMHO). I've retested against master, to no avail.

- I'm authenticating against the OVH API using this [rake task](https://gist.github.com/tiennou/21a91b0e0fa7f8a65e0998d05c978855), because it gets painful to do copy-pasting of URLs. I got the `app-key` while following the setup guide.
- I can push/preview all defined domains successfully (which means my zones are now up-to-date), it only fails when I'm asking for a specific domain, using the commands as I provided them above, by substituting a defined domain. Of note, requesting `one-of-my.domains` (ie. a really undefined one) results when previewing/pushing as `Done. 0 corrections.`.
- I'm using some Javascript-ing to handle OVH ALLDOM aliasing, but those end-up calling `D()`. I've also checked straight `D()` domains, to the same effect.
- I've checked that commenting all `D()` stanzas except one, and asking for that specific one results in the same thing.
- The OVH API list zone API has the failing domains (though I'm connected via my main account and not through my app-key identity, I haven't been able to find out if that's even possible).
- I don't make typoes (sorry, couldn't help) 😜.

If you provide me with source-code coordinates, I can trying dumping some things to see what's happening, though I'll try to see if I can manage on my own.

**Comment by masterzen** [comment]
@tiennou,

The code is very simple. When trying to apply the corrections we [end here](https://github.com/StackExchange/dnscontrol/blob/master/providers/ovh/ovhProvider.go#L88) where the error you see is raised.
If you add some `log.Print` around there, you could list the content of `zones`, and see what OVH returned to `dnscontrol`.
This `zones` array is populated in [ovh/protocol.go](https://github.com/StackExchange/dnscontrol/blob/master/providers/ovh/protocol.go#L16), and it is coming straight out of OVH API.
If the OVH API doesn't return the domain in question then that means there's something different between the account you use with `dnscontrol` and your OVH console account (by definition the OVH console has all access granted I think).
You could also do a simple [`curl` with your ovh tokens](https://docs.ovh.com/gb/en/customer/first-steps-with-ovh-api/#carrying-out-a-request).
To rule out a permission issue, you can generate a consumer key with `GET /*` and `PUT /*` permissions and see if that fixes or not the issue.

**Comment by tiennou** [comment]
I think I found the issue (see the 2nd commit in my branch above). AFAICS, when `--domains` is used, when `DetermineNameservers` happens `NumberOfNameservers` is 0, so the provider/driver's `GetNameservers` call is skipped, thus the API is never queried for the list of valid domains. When `--domains` is not used, `NumbersOfNameservers` is `-1` thus the call happen and the API is queried.

I'm not sure the fix is correct, since it seems the bug is caused by this `-1`/`0` change.

**Comment by masterzen** [comment]
@tiennou,

Good catch!

Indeed the `zones` are cached in the OVH provider in `GetNameservers`, so if this is not called at all, then we don't have any zones. I will make sure it is called during provider registration.

Though on a yesterday's master dnscontrol, `GetNameservers` is called when using `--domains`:
~~~
% ./dnscontrol preview --domains dnscontroltest.ovh
******************** Domain: dnscontroltest.ovh
----- Getting nameservers from: ovh
----- DNS Provider: ovh...3 corrections
#1: MODIFY NS dnscontroltest.ovh: (dns19.ovh.net ttl=300) -> (dns19.ovh.net. ttl=300)
#2: MODIFY NS dnscontroltest.ovh: (ns19.ovh.net ttl=300) -> (ns19.ovh.net. ttl=300)
#3: REFRESH zone dnscontroltest.ovh
----- Registrar: none...0 corrections
Done. 3 corrections.
~~~

This number of nameservers comes from your declaration of `DnsProvider` in your `dnsconfig.js` as in:
~~~
var ovh = NewDnsProvider("ovh", "OVH");
...
D("dnscontroltest.ovh", REG_NONE, DnsProvider(ovh),
...
)
~~~

If `nsCount` is not provided, it defaults to `-1` which means all nameservers will be taken into account. But if `NumberOfNameservers` is `0`, then indeed we won't use any nameserver, and then the OVH provider `GetNameservers` will never be called (and `zones` will be empty).

I will fix this, but I fail to see how the `NumberOfNameservers` can become `0` in the `--domains` case. Looking at [helpers.js#50](https://github.com/StackExchange/dnscontrol/blob/master/pkg/js/helpers.js#L150), there's no way it can become `0` unless `0` has been passed in `nsCount`.

Can you add more debug around there to see what could cause `0` to be passed in?

Thanks!

**Comment by masterzen** [comment]
@tiennou, I pushed a fix in PR #417. 
Can you try it?

**Comment by tiennou** [comment]
> Can you add more debug around there to see what could cause 0 to be passed in?

Well, that's easy : I'm the one passing that 😉. I'm still in the process of migrating, so for now I'm just setting up the OVH-managed zones to be "copies" of my Plesk/Bind based "production" ones, so I want to ignore changes to the nameservers themselves.

**Comment by masterzen** [comment]
@tiennou,

> Well, that's easy : I'm the one passing that 😉

Ah, that was the missing piece of information for reproducing the problem locally :)
Indeed I never thought about running with 0 nameservers :)

Anyhow, thanks for confirming #417 is fixing the issue for you.

**Comment by masterzen** [comment]
@tlimoncelli this can be closed, it was fixed by #417.
